src/main/java/org/inferred/freebuilder/processor/property/BiMapProperty.java

```diff
@@ -0,0 +1,406 @@
+      Optional<TypeMirror> unboxedValueType = maybeUnbox(valueType, config.getTypes());
+      boolean overridesPutMethod = hasPutMethodOverride(
+          config, unboxedKeyType.orElse(keyType), unboxedValueType.orElse(valueType));
+
+      FunctionalType mutatorType = functionalTypeAcceptedByMethod(
+          config.getBuilder(),
+          mutator(property),
+          consumer(biMap(keyType, valueType, config.getElements(), config.getTypes())),
+          config.getElements(),
+          config.getTypes()
+      );
+
+      return Optional.of(new BiMapProperty(
+          config.getDatatype(),
+          property,
+          overridesPutMethod,
+          keyType,
+          unboxedKeyType,
+          valueType,
+          unboxedValueType,
+          mutatorType
+      ));
+    }
+
+    private static boolean hasPutMethodOverride(
+        Config config, TypeMirror keyType, TypeMirror valueType
+    ) {
+      return overrides(
+          config.getBuilder(),
+          config.getTypes(),
+          putMethod(config.getProperty()),
+          keyType,
+          valueType
+      );
+    }
+
+    private static TypeMirror biMap(
+        TypeMirror keyType,
+        TypeMirror valueType,
+        Elements elements,
+        Types types
+    ) {
+      TypeElement mapType = elements.getTypeElement(BiMap.class.getName());
+      return types.getDeclaredType(mapType, keyType, valueType);
+    }
+  }
+
+  private final boolean overridesPutMethod;
+  private final TypeMirror keyType;
+  private final Optional<TypeMirror> unboxedKeyType;
+  private final TypeMirror valueType;
+  private final Optional<TypeMirror> unboxedValueType;
+  private final FunctionalType mutatorType;
+
+  BiMapProperty(
+      Datatype datatype,
+      Property property,
+      boolean overridesPutMethod,
+      TypeMirror keyType,
+      Optional<TypeMirror> unboxedKeyType,
+      TypeMirror valueType,
+      Optional<TypeMirror> unboxedValueType,
+      FunctionalType mutatorType
+  ) {
+    super(datatype, property);
+    this.overridesPutMethod = overridesPutMethod;
+    this.keyType = keyType;
+    this.unboxedKeyType = unboxedKeyType;
+    this.valueType = valueType;
+    this.unboxedValueType = unboxedValueType;
+    this.mutatorType = mutatorType;
+  }
+
+  @Override
+  public void addValueFieldDeclaration(SourceBuilder code) {
+    code.addLine(
+        "private final %s<%s, %s> %s;",
+        ImmutableBiMap.class,
+        keyType,
+        valueType,
+        property.getField()
+    );
+  }
+
+  @Override
+  public void addBuilderFieldDeclaration(SourceBuilder code) {
+    code.addLine(
+        "private final %1$s<%2$s, %3$s> %4$s = %1$s.create();",
+        HashBiMap.class,
+        keyType,
+        valueType,
+        property.getField()
+    );
+  }
+
+  @Override
+  public void addBuilderFieldAccessors(SourceBuilder code) {
+    addPut(code);
+    addPutAll(code);
+    addRemoveKey(code);
+    addRemoveValue(code);
```

Comment:
This may have been misguided but since BiMap's are 1-1, you may want to make modifications based on the value and not the key.

Reply:
Seems reasonable. You _could_ do `.mutateX(m -> m.inverse().remove(value))` so it really comes down to convenience.